Sources/Bodega/StorageEngine.swift

```diff
@@ -0,0 +1,48 @@
+import Foundation
+
+/// A `StorageEngine` represents an underlying data store mechanism for saving and persisting data.
+///
+/// A ``StorageEngine`` is a construct you can build that plugs into ``ObjectStorage``
+/// to use for persisting data.
+///
+/// This library has two implementations of `StorageEngine`, ``DiskStorageEngine`` and ``SQLiteStorageEngine``.
+/// Both of these can serve as inspiration if you have your own persistence mechanism (such as Realm, CoreData, etc).
+///
+/// `DiskStorageEngine` takes `Data` and saves it to disk using file system operations.
+/// `SQLiteStorageEngine` takes `Data` and saves it to an SQLite database under the hood.
+///
+/// If you have your own way of storing data then you can refer to `DiskStorageEngine` and `SQLiteStorageEngine`
+/// for inspiration, but all you need to do is conform to the `StorageEngine` protocol
+/// and initialize `ObjectStorage` with that storage.
+public protocol StorageEngine: Actor {
+    var directory: FileManager.Directory { get }
+
+    func write(_ data: Data, key: CacheKey) throws
+    func write(_ dataAndKeys: [(key: CacheKey, data: Data)]) throws
+
+    func read(key: CacheKey) -> Data?
+    func read(keys: [CacheKey]) -> [Data]
+    func readDataAndKeys(keys: [CacheKey]) -> [(key: CacheKey, data: Data)]
+    func readAllData() -> [Data]
+    func readAllDataAndKeys() -> [(key: CacheKey, data: Data)]
```

Comment:
seems like a good default impl. based off the concrete instances might be
```swift
    public func readAllDataAndKeys() -> [(key: CacheKey, data: Data)] {
        let allKeys = self.allKeys()
        return self.readDataAndKeys(keys: allKeys)
    }
```

these are just suggestions though! if you feel default impls aren't the way to go that makes sense also :D

Reply:
These are really great suggestions! I love using default implementations to lessen the work and remove duplicate code. I don't foresee any issues doing it here, and found a couple more candidates for default implementations. 🙂